File: frontend/lib/services/osm_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:geocoding/geocoding.dart' as gc;

class OSMService {
  static final OSMService _instance = OSMService._internal();
  factory OSMService() => _instance;
  OSMService._internal();

  // Nominatim API endpoint (free, open source)
  static const String _nominatimUrl = 'https://nominatim.openstreetmap.org';
  final Map<String, String> _placeNameCache = {};

  /// Get place name from coordinates using reverse geocoding
  Future<String> getPlaceNameFromCoordinates(double latitude, double longitude) async {
    final cacheKey = '$latitude,$longitude';
    
    // Check cache first
    if (_placeNameCache.containsKey(cacheKey)) {
      return _placeNameCache[cacheKey]!;
    }

    try {
      // Try using built-in geocoding first (faster)
      final placemarks = await gc.placemarkFromCoordinates(latitude, longitude);
      if (placemarks.isNotEmpty) {
        final placemark = placemarks.first;
        String placeName = placemark.name ?? '';
        if (placemark.locality != null && placemark.locality!.isNotEmpty) {
          placeName = '${placemark.locality}, ${placemark.administrativeArea}';
        }
        if (placeName.isEmpty && placemark.street != null) {
          placeName = placemark.street ?? 'Unknown Location';
        }
        
        _placeNameCache[cacheKey] = placeName;
        return placeName;
      }
    } catch (e) {
      print('Built-in geocoding failed: $e');
    }

    // Fallback to Nominatim API
    try {
      final uri = Uri.parse(
        '$_nominatimUrl/reverse?format=json&lat=$latitude&lon=$longitude&zoom=18&addressdetails=1'
      );
      
      final response = await http.get(
        uri,
        headers: {'User-Agent': 'Net-Fence-AI/1.0'},
      ).timeout(const Duration(seconds: 10));

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        
        // Extract place name from response
        String placeName = data['name'] ?? 'Unknown Location';
        
        if (data['address'] != null) {
          final address = data['address'];
          final parts = <String>[];
          
          if (address['road'] != null) parts.add(address['road']);
          if (address['suburb'] != null) parts.add(address['suburb']);
          if (address['city'] != null) parts.add(address['city']);
          if (address['country'] != null) parts.add(address['country']);
          
          if (parts.isNotEmpty) {
            placeName = parts.join(', ');
          }
        }
        
        _placeNameCache[cacheKey] = placeName;
        return placeName;
      }
    } catch (e) {
      print('Nominatim reverse geocoding failed: $e');
    }

    // Return coordinates as fallback
    final fallback = '${latitude.toStringAsFixed(4)}, ${longitude.toStringAsFixed(4)}';
    _placeNameCache[cacheKey] = fallback;
    return fallback;
  }

  /// Get coordinates from place name
  Future<Map<String, double>?> getCoordinatesFromPlaceName(String placeName) async {
    try {
      final locations = await gc.locationFromAddress(placeName);
      if (locations.isNotEmpty) {
        return {
          'latitude': locations.first.latitude,
          'longitude': locations.first.longitude,
        };
      }
    } catch (e) {
      print('Built-in geocoding for place name failed: $e');
    }

    // Fallback to Nominatim
    try {
      final uri = Uri.parse(
        '$_nominatimUrl/search?format=json&q=${Uri.encodeComponent(placeName)}&limit=1'
      );
      
      final response = await http.get(
        uri,
        headers: {'User-Agent': 'Net-Fence-AI/1.0'},
      ).timeout(const Duration(seconds: 10));

      if (response.statusCode == 200) {
        final results = jsonDecode(response.body);
        if (results.isNotEmpty) {
          return {
            'latitude': double.parse(results[0]['lat'].toString()),
            'longitude': double.parse(results[0]['lon'].toString()),
          };
        }
      }
    } catch (e) {
      print('Nominatim geocoding failed: $e');
    }

    return null;
  }

  /// Get detailed location info from coordinates
  Future<Map<String, dynamic>> getLocationDetails(double latitude, double longitude) async {
    try {
      final uri = Uri.parse(
        '$_nominatimUrl/reverse?format=json&lat=$latitude&lon=$longitude&zoom=18&addressdetails=1&extratags=1&namedetails=1'
      );
      
      final response = await http.get(
        uri,
        headers: {'User-Agent': 'Net-Fence-AI/1.0'},
      ).timeout(const Duration(seconds: 10));

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        
        return {
          'name': data['name'] ?? 'Unknown',
          'address': data['address'] ?? {},
          'display_name': data['display_name'] ?? '',
          'osm_id': data['osm_id'],
          'osm_type': data['osm_type'],
          'importance': data['importance'] ?? 0.0,
          'place_rank': data['place_rank'] ?? 0,
        };
      }
    } catch (e) {
      print('Error getting location details: $e');
    }

    return {};
  }

  /// Clear the cache
  void clearCache() {
    _placeNameCache.clear();
  }

  /// Get cache statistics
  Map<String, int> getCacheStats() {
    return {
      'cached_locations': _placeNameCache.length,
    };
  }
}

File: backend/services/osm_service.py
import requests
from functools import lru_cache
from typing import Dict, Tuple
import logging

logger = logging.getLogger(__name__)

class OSMService:
    """Service for OpenStreetMap/Nominatim integration"""
    
    NOMINATIM_BASE_URL = "https://nominatim.openstreetmap.org"
    
    @staticmethod
    @lru_cache(maxsize=256)
    def get_place_name(latitude: float, longitude: float) -> str:
        """
        Get place name from coordinates using reverse geocoding via Nominatim.
        Results are cached for performance.
        """
        try:
            url = f"{OSMService.NOMINATIM_BASE_URL}/reverse"
            params = {
                'format': 'json',
                'lat': latitude,
                'lon': longitude,
                'zoom': 18,
                'addressdetails': 1
            }
            headers = {'User-Agent': 'Net-Fence-AI/1.0'}
            
            response = requests.get(url, params=params, headers=headers, timeout=5)
            response.raise_for_status()
            
            data = response.json()
            
            # Extract place name from response
            if 'address' in data:
                address = data['address']
                parts = []
                
                # Build address parts in order of priority
                if 'road' in address:
                    parts.append(address['road'])
                if 'city' in address:
                    parts.append(address['city'])
                elif 'suburb' in address:
                    parts.append(address['suburb'])
                if 'country' in address and address['country'] != 'India':
                    parts.append(address['country'])
                
                if parts:
                    return ', '.join(parts)
            
            # Fallback to display name
            if 'name' in data:
                return data['name']
            
            return f"{latitude:.4f}, {longitude:.4f}"
            
        except Exception as e:
            logger.warning(f"Error getting place name for {latitude}, {longitude}: {e}")
            return f"{latitude:.4f}, {longitude:.4f}"
    
    @staticmethod
    def get_location_details(latitude: float, longitude: float) -> Dict:
        """
        Get detailed location information from coordinates.
        """
        try:
            url = f"{OSMService.NOMINATIM_BASE_URL}/reverse"
            params = {
                'format': 'json',
                'lat': latitude,
                'lon': longitude,
                'zoom': 18,
                'addressdetails': 1,
                'extratags': 1
            }
            headers = {'User-Agent': 'Net-Fence-AI/1.0'}
            
            response = requests.get(url, params=params, headers=headers, timeout=5)
            response.raise_for_status()
            
            data = response.json()
            
            return {
                'name': data.get('name', 'Unknown'),
                'address': data.get('address', {}),
                'display_name': data.get('display_name', ''),
                'osm_id': data.get('osm_id'),
                'osm_type': data.get('osm_type'),
                'latitude': latitude,
                'longitude': longitude,
            }
            
        except Exception as e:
            logger.warning(f"Error getting location details for {latitude}, {longitude}: {e}")
            return {
                'name': 'Unknown Location',
                'latitude': latitude,
                'longitude': longitude,
            }
    
    @staticmethod
    def clear_cache():
        """Clear the geocoding cache"""
        OSMService.get_place_name.cache_clear()
